enhancements/dns/nodeservice-etc-hosts-injection.md

```diff
@@ -0,0 +1,422 @@
+- Filter out old managed entries before adding updated ones
+- Only update `/etc/hosts` if valid service IPs are resolved
+
+#### Environment Variable Considerations
+
+The implementation uses a Kubernetes DaemonSet environment variable named `SERVICES` to pass a list of service hostnames to the DNS node resolver pods. The `update-node-resolver.sh` script then dynamically looks up the IP addresses for these services. This approach has several important implications:
+
+**Advantages:**
+- Simple and reliable mechanism built into Kubernetes
+- No need for shared storage or complex synchronization
+- Standard Kubernetes primitives for configuration management
+- **Service IP changes don't require pod restarts**: The script detects and updates IPs dynamically via 60-second polling
+- Small footprint: only service names in `service-name.namespace.svc` format are stored, not IP addresses
+- Efficient: ~1.4KB for 20 services, minimal impact on environment variable limits
+- Flexible format: supports comma-separated, space-separated, or mixed delimiters
+
+**Limitations:**
+- **Service list changes require pod restarts**: When services are added to or removed from the `nodeServices` list in the DNS CR, the operator must update the DaemonSet `SERVICES` environment variable, triggering a rolling restart of all node resolver pods
+  - Rolling restart respects PodDisruptionBudgets and ensures gradual propagation
+  - Brief window where some nodes may be tracking different service lists during rolling update
+- **Script-based IP lookup**: The `update-node-resolver.sh` script must:
+  - Run in a continuous 60-second polling loop
+  - Resolve service IPs via DNS using `dig` command (tries IPv4, IPv6, and TCP fallback)
+  - Handle services that don't exist or have no ClusterIP
+  - Preserve stale entries if DNS resolution fails temporarily (doesn't remove working entries during outages)
+- **Size constraints**: Total environment variable size for a pod is limited (typically 128KB)
+  - 20 services limit provides safety margin
+  - Each service name in `service-name.namespace.svc` format is approximately 30-70 bytes
+  - Leaves ample space for other environment variables
+- **Visibility**: Service names are visible in pod specs and environment variables (not sensitive data)
+
+**Update Propagation:**
+
+**For service list changes** (adding/removing services from `nodeServices`):
+- The operator updates the DaemonSet `SERVICES` environment variable
+- This triggers a rolling update with the following characteristics:
+  - Gradual rollout across all nodes (default: 1 pod at a time)
+  - Respects node scheduling constraints and taints
+  - Typically completes within minutes for most clusters
+  - Different nodes temporarily track different service lists during the rollout
+
+**For service IP changes** (ClusterIP of a tracked service changes):
+- The `update-node-resolver.sh` script detects the IP change during its next 60-second polling cycle
```

Comment:
Once the DNS infrastructure is up and running, does it make sense to continue to have these host entries as a fallback? CoreDNS has its own caching doesn't it, so once it has fetched each of these, it should be able to serve the other host components?